src/RelayProvider.ts

```diff
@@ -0,0 +1,270 @@
+        }
+
+        if (
+            relayHubValue === undefined ||
+            relayHubValue === null ||
+            relayHubValue === constants.AddressZero
+        ) {
+            relayHubValue = config.relayHubAddress;
+        }
+
+        if (
+            onlyPreferredRelaysValue === undefined ||
+            onlyPreferredRelaysValue === null
+        ) {
+            onlyPreferredRelaysValue = config.onlyPreferredRelays;
+        }
+
+        if (gasToSend !== undefined && gasToSend !== null) {
+            gasToSend = hexValue(gasToSend);
+        }
+
+        /**
+         * When using a RelayProvider, and the gas field is not manually set, then the original provider within calculates a ridiculously high gas.
+         * In order to avoid this, we created a new field called forceGas, which is optional.
+         * If the user wants to manually set the gas, then she can put a value in the forceGas field, otherwise it will be
+         * automatically estimated (correctly) by the RelayClient.
+         * The reason why we added this new forceGas field is because at this point, by reading the gas field,
+         * we cannot differentiate between a user-entered gas value from the auto-calculated (and a very poor calculation)
+         * value that comes from the original provider
+         */
+
+        const fullEnvelopingRequest: UserDefinedEnvelopingRequest = {
+            relayData: {
+                ...envelopingRequest.relayData,
+                callForwarder: callForwarderValue,
+            }, 
+            request: {
+                ...envelopingRequest.request,
+                relayHub: relayHubValue,
+                gas: gasToSend,
+            }
+        }
+
+        const fullRequestConfig: RequestConfig = {
+            ...requestConfig,
+            onlyPreferredRelays: onlyPreferredRelaysValue
+        }
+
+        return this.executeRelayTransaction(fullEnvelopingRequest, fullRequestConfig);
+    }
+
+    private async _getAccounts(method: string, params: Array<{ requestConfig: RequestConfig, envelopingTx: UserDefinedEnvelopingRequest }>) {
+
+        const response = await this.jsonRpcProvider.send(method, params) as { result: unknown[] };
+
+        if (response != null && Array.isArray(response.result)) {
+            const { chainId } = await this.getNetwork();
+            const accountManager = new AccountManager(this, chainId);
+
+            const ephemeralAccounts =
+                accountManager.getAccounts();
+
+            response.result = response.result.concat(ephemeralAccounts);
+        }
+
+        return response;
+    }
+
+    override send(method: string, params: Array<{ requestConfig: RequestConfig, envelopingTx: UserDefinedEnvelopingRequest }>): Promise<unknown> {
+
+        const [ envelopingData ] = params;
+        if(!envelopingData) {
+            throw new Error('No data to send!');
+        }
+
+        const { requestConfig, envelopingTx: { request } } = envelopingData;
```

Comment:
I'm not sure that this is going to work for transactions that are not intended to be relayed. If we try to extract properties from undefined properties, it raises an error, and I believe that neither `envelopingData` or `envelopingTx` is going to be defined in that case.

Reply:
I have changed the function signature to be more flexible